lang/udf_cache.hh

```diff
@@ -0,0 +1,154 @@
+
+/*
+ * SPDX-License-Identifier: (AGPL-3.0-or-later and Apache-2.0)
+ */
+
+#pragma once
+
+#include "cql3/functions/function_name.hh"
+#include "utils/overloaded_functor.hh"
+#include "wasmtime.hh"
+#include "seastar/core/shared_ptr.hh"
+#include "seastar/core/scheduling.hh"
+#include "utils/loading_cache.hh"
+#include "utils/hash.hh"
+#include "utils/overloaded_functor.hh"
+#include <chrono>
+#include <cstdint>
+#include <list>
+
+extern logging::logger wasm_logger;
+
+namespace wasm {
+
+struct udf_instance {
+    wasmtime::Store store;
+    wasmtime::Instance instance;
+    wasmtime::Func func;
+    wasmtime::Memory memory;
+};
+
+// For each UDF full name and a scheduling group, we store a future that is ready
+// when no other task uses the same UDF instance. When an instance is retrieved
+// from the cache, the future is replaced with a new one that becomes ready when
+// the instance is returned. To avoid conflicts between tasks, each time an 
+// instance is requested, the future associated with it is removed from the cache
+// and a new one is added. When the current instance user is done with it, the
+// future that was associated with it is becomes ready.
+// This way, an instance stays in the cache as long as it isn't used by any task
+// and if there are tasks that want to use it, they get it immediately after
+// the previous user is finished.
+class udf_cache {
+public:
+    struct stats {
+        uint64_t cache_hits = 0;
+        uint64_t cache_misses = 0;
+        uint64_t cache_blocks = 0;
+    };
+
+    static stats& shard_stats() {
+        static thread_local stats _stats;
+        return _stats;
+    }
+
+private:
+    using cache_key_type = std::pair<seastar::scheduling_group, cql3::functions::function_name>;
+    using cache_entry_type = seastar::future<udf_instance>;
+    struct udf_cache_entry_size {
+        size_t operator()(cache_entry_type& val) {
+            if (!val.available()) {
+                // the instance is currently being used
+                return 0;
+            } else {
+                auto&& inst = val.get0();
+                uint32_t size = inst.memory.size(inst.store);
+                val = seastar::make_ready_future<udf_instance>(std::move(inst));
+                return size;
+            }
+        }
+    };
+    using cache_type = utils::loading_cache<cache_key_type, cache_entry_type, 1, utils::loading_cache_reload_enabled::no, udf_cache_entry_size, utils::tuple_hash, std::equal_to<cache_key_type>>;
+
+public:
+    using key_type = cql3::functions::function_name;
+    using value_type = udf_instance;
+
+private:
+    cache_type _cache;
+    size_t _max_size;
+
+public:
+    udf_cache(size_t size)
+        : _cache(size, std::chrono::seconds(10), wasm_logger)
+        , _max_size(size)
+    {}
+
+    template <typename LoadFunc>
+    requires std::is_invocable_r_v<value_type, LoadFunc>
+    std::pair<value_type, seastar::promise<value_type>> get(const key_type& key, LoadFunc&& load) {
```

Comment:
I see you have a specialized wrapper here, so regardless of loading_cache issues, you can move LoadFunc here and absolve the caller from having to supply it. It can supply the extra parameters LoadFunc needs as regular parameters.

Reply:
Why does it return both a value_type and a promise<value_type>?

Reply:
Ah, is that for returning the instance? I think it should just call put(), future/promise isn't needed for synchronous communication.

Reply:
> I see you have a specialized wrapper here, so regardless of loading_cache issues, you can move LoadFunc here and absolve the caller from having to supply it. It can supply the extra parameters LoadFunc needs as regular parameters.

I'll do that

> Ah, is that for returning the instance? I think it should just call put(), future/promise isn't needed for synchronous communication.

Yes, it's for returning. We don't have to use the futures now, but when we add pre-empting to udf calls, I think we will need them